Explicit instantiation after template specialization warning

clang in c++98 mode issues the following warning [1]:
```
/.../dealii/source/lac/trilinos_sparse_matrix.cc:1741:17: warning: explicit instantiation
of 'reinit<dealii::DynamicSparsityPattern>' that occurs after an explicit specialization
will be ignored (C++11 extension) [-Wc++11-extensions]

  SparseMatrix::reinit (const Epetra_Map &,
                ^
```

This might be a subtle problem we should address.

[1] http://cdash.kyomu.43-1.org/viewBuildError.php?type=1&buildid=6058

--- a/include/deal.II/lac/trilinos_sparse_matrix.h
+++ b/include/deal.II/lac/trilinos_sparse_matrix.h
@@ -57,6 +57,7 @@ template <typename MatrixType> class BlockMatrixBase;
 
 template <typename number> class SparseMatrix;
 class SparsityPattern;
+class DynamicSparsityPattern;
 
 
 namespace TrilinosWrappers
@@ -2944,6 +2945,14 @@ namespace TrilinosWrappers
     reinit (row_map, col_map, sparsity_pattern, exchange_data);
   }
 
+
+  // declare the existence of an explicit specialization
+  template <>
+  void
+  SparseMatrix::reinit (const Epetra_Map    &input_row_map,
+                        const Epetra_Map    &input_col_map,
+                        const DynamicSparsityPattern &sparsity_pattern,
+                        const bool           exchange_data);
 
 
   template <typename number>

--- a/source/lac/trilinos_sparse_matrix.cc
+++ b/source/lac/trilinos_sparse_matrix.cc
@@ -1737,12 +1737,6 @@ namespace TrilinosWrappers
                         const dealii::SparsityPattern &,
                         const bool);
 
-  template void
-  SparseMatrix::reinit (const Epetra_Map &,
-                        const Epetra_Map &,
-                        const DynamicSparsityPattern &,
-                        const bool);
-
 }
 
 DEAL_II_NAMESPACE_CLOSE

